File: codigo/soluciones/kat/10kindsofpeople.py
def bfs(i, j, m, ids, r, c, id):
    if ids[i][j] != 0:
        return
    q = []
    q.append((i, j))
    ids[i][j] = id
    while q:
        i, j = q.pop()
        for d in [-1, 1]:
            if 0 <= i + d < r and m[i + d][j] == m[i][j] and ids[i + d][j] == 0:
                q.append((i + d, j))
                ids[i + d][j] = id
            if 0 <= j + d < c and m[i][j + d] == m[i][j] and ids[i][j + d] == 0:
                q.append((i, j + d))
                ids[i][j + d] = id
r, c = map(int, input().split())
m = [input().strip() for _ in range(r)]
ids = [[0 for _ in range(c)] for _ in range(r)]
id = 0
for i in range(r):
    for j in range(c):
        if ids[i][j] != 0:
            continue
        id += 1
        bfs(i, j, m, ids, r, c, id)
n = int(input())
for _ in range(n):
    r1, c1, r2, c2 = map(int, input().split())
    r1 -= 1
    c1 -= 1
    r2 -= 1
    c2 -= 1
    if (ids[r1][c1] == ids[r2][c2]):
        print("binary" if m[r1][c1] == '0' else "decimal")
    else:
        print("neither")

File: codigo/soluciones/kat/10kindsofpeople.java
import java.io.*;
import java.util.*;
class Tenkindsofpeople {
    private void dfs(int i, int j, char[][] m, int[][] ids, int r, int c, int id) {
    Deque<int[]> stack = new ArrayDeque<>();
    stack.push(new int[]{i, j});
    ids[i][j] = id;
    while (!stack.isEmpty()) {
        int[] cur = stack.pop();
        i = cur[0];
        j = cur[1];
        for (int d = -1; d <= 1; d += 2) {
            int ni = i + d, nj = j;
            if (ni >= 0 && ni < r && nj >= 0 && nj < c && m[ni][nj] == m[i][j] && ids[ni][nj] == 0) {
                ids[ni][nj] = id;
                stack.push(new int[]{ni, nj});
            }
            ni = i;
            nj = j + d;
            if (ni >= 0 && ni < r && nj >= 0 && nj < c && m[ni][nj] == m[i][j] && ids[ni][nj] == 0) {
                ids[ni][nj] = id;
                stack.push(new int[]{ni, nj});
            }
        }
    }
}
    private void _main() throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int r = Integer.parseInt(st.nextToken()), c = Integer.parseInt(st.nextToken());
        char[][] m = new char[r][c];
        for (int i = 0; i < r; ++i) {
            String s = br.readLine();
            for (int j = 0; j < c; ++j) {
                m[i][j] = s.charAt(j);
            }
        }
        int[][] ids = new int[r][c];
        int id = 0;
        for (int i = 0; i < r; ++i) {
            for (int j = 0; j < c; ++j) {
                if (ids[i][j] != 0) continue;
                dfs(i, j, m, ids, r, c, ++id);
            }
        }
        int n = Integer.parseInt(br.readLine());
        StringBuilder out = new StringBuilder();
        while (n > 0) {
            --n;
            st = new StringTokenizer(br.readLine());
            int r1 = Integer.parseInt(st.nextToken()), c1 = Integer.parseInt(st.nextToken()), r2 = Integer.parseInt(st.nextToken()), c2 = Integer.parseInt(st.nextToken());
            --r1;
            --c1;
            --r2;
            --c2;
            if (ids[r1][c1] == ids[r2][c2]) {
                out.append(m[r1][c1] == '0' ? "binary" : "decimal");
            }else {
                out.append("neither");
            }
            out.append("\n");
        }
        bw.write(out.toString());
        bw.flush();
        bw.close();
    }
    public static void main(String[] args) throws IOException {
        (new Tenkindsofpeople())._main();
    }
}
